### PR TITLE
Tests for Simple-Data-Grapher

### PR DESCRIPTION
**Pull request by IshaGupta18**
Here are a few of the tests I have written for `csvfiles_controller.rb` file and I will add more as and when required.
Since I have done most of my testing (logical and UI) in Simple-Data-Grapher as well, I think these will be good to go for now.
Also, this gives more confidence to #5993 which can be merged soon, now that we have tests to back it up.
@jywarren @gauravano @sagarpreet-chadha @namangupta01 kindly review this and let me know what else can be done!

### CODE REVIEW

**Comment by IshaGupta18** [comment]
PS: I would be fixing the linting soon!

**Comment by IshaGupta18** [comment]
Here only the test file bears important, rest all is from #5993 https://github.com/publiclab/plots2/pull/6104/files#diff-b6813d8a7c58e0cf4118a4a9e17873cd

File: app/views/csvfiles/user_files.html.erb [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+                    <p> No Description </p>
+                <% end %>
+            </td>
+            <% count += 1 %>
+        </tr>
+    <% end %>
+</table>
+
+<style>
+.main_heading_data{
```
Comment (IgorWilbert) [inline app/views/csvfiles/user_files.html.erb]: A separate file for these css styles would be great I think

File: test/functional/csvfiles_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		get :prev_files, params: {uid: 2}
+		assert_response :success
+	end
+
+	test 'should not get user files if not logged in'  do
+		get :user_files, params: {id: 2}
+		assert_redirected_to '/login'
+	end
+
+	test 'should save graph object' do
```
Comment (IgorWilbert) [inline test/functional/csvfiles_controller_test.rb]: Can you think of other unhappy path tests for this? For example, is there any circumstance in which the save/fetch graph object _should_ fail? (e.g. user not logged in)

**Comment by IgorWilbert** [comment]
Thank you @IshaGupta18 ! I left some comments, I hope they are helpful :)

File: test/functional/csvfiles_controller_test.rb (hunk 2 above)
Comment (IshaGupta18) [reply test/functional/csvfiles_controller_test.rb]: For fetching, maybe yes! But for saving, I have checked for the login redirect!

File: app/views/csvfiles/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+<script src="https://cdnjs.cloudflare.com/ajax/libs/xlsx/0.14.3/xlsx.full.min.js"></script>
+<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.8.2/css/all.min.css">
+<%= javascript_include_tag('/lib/simple-data-grapher/dist/PublicLab.Grapher.js')%>
+<%= stylesheet_link_tag '/lib/simple-data-grapher/examples/upload_file.css'%>
+<div id="first"></div>
+<script>
+    var headerContainer = document.getElementsByClassName("body-container")[0];
+    SimpleDataGrapherObject = new SimpleDataGrapher("first");
+    var value = '<%= current_user %>';
+    <% if current_user %>
```
[the remark this answers is not in the document]
Comment (IshaGupta18) [reply app/views/csvfiles/new.html.erb]: Actually, this file is very much like the [ rich.html.erb ](https://github.com/publiclab/plots2/blob/master/app/views/editor/rich.html.erb), which integrates the PL Editor in the same way as Simple-Data-Grapher, that is why it's required here!

**Comment by jywarren** [comment]
Hi!!! Does this include all the commits from https://github.com/publiclab/plots2/pull/5993? If so, I'd prefer to merge this instead of that, to do it all in one. That way the tests and functionality are linked!

**Comment by IshaGupta18** [comment]
Yes this does, however I have a couple of bugs to fix in this one, then
it's good for merge! Thanks a lot!

File: app/controllers/editor_controller.rb [hunk 4]
```diff
@@ -20,6 +20,11 @@ def editor
   end
 
   def post
+    @graphobject = ""
+    if params[:id] && params[:uid]
+      @graphobject = [params[:id],params[:uid]]
+    end
+    # byebug
```
Comment (jywarren) [inline app/controllers/editor_controller.rb]: Hm, is this required?

File: app/views/editor/rich.html.erb [hunk 5]
```diff
@@ -204,11 +204,19 @@
...
-      <div class="ple-module-content col-lg-9">
-        <textarea id="text-input" class="ple-textarea form-control"><% if @node && @node.latest && @node.latest.body_rich %><%= @node.latest.body %><% else %><%= params[:body] %><% end %></textarea>
-      </div>
+      <% if @graphobject=="" %>
+        <div class="ple-module-content col-lg-9">
+          <textarea id="text-input" class="ple-textarea form-control"><% if @node && @node.latest && @node.latest.body_rich %><%= @node.latest.body %><% else %><%= params[:body] %><% end %></textarea>
+        </div>
+      <% else %>
+        <div class="ple-module-content col-lg-9">
+          <% x = "Power Tag: simple-data-grapher: i/" + @graphobject[0] + "/" + @graphobject[1] %>
```
Comment (jywarren) [inline app/views/editor/rich.html.erb]: Hm, can you walk me through exactly what's happening here, or link to the previous PR where it's explained? Thanks!

**Comment by IshaGupta18** [comment]
Yes, I am will write a comment explaining the entire flow and also why both
these code snippets highlighted by you are necessary!

**Comment by IshaGupta18** [comment]
Okay so here is the flow of things, a part of which I will be borrowing from my comment in https://github.com/publiclab/simple-data-grapher/issues/65.

The User clicks on "Publish as a Research Note" button 
image
When this happens, as suggested by @sagarpreet-chadha and @gauravano, I save a hash containing the data I will be needing to plot the same graph again. This is done by sending an AJAX to the controller action in `csvfiles_controller.rb` , which creates a new csvfiles object, and returns just the `id` and `uid` of this object.
image

We need this `id` and `uid` in order to create the `Power Tag` for `simple-data-grapher`.

Once we get this, we give the control to the `editor_controller.rb` file by doing this:
image
We have passed the `id` and `uid` in params here, which is received by the `post` function (which has the control now)
image

The global variable `@graphobject` is available to the corresponding view that `post` function fetches ie `rich.html.erb`

There, we just use a simple if condition, to check if we need to insert a power tag for simple-data-grapher or not.

image

image


The remaining flow is as described in [this comment](https://github.com/publiclab/simple-data-grapher/issues/65#issuecomment-510342825)
 ie. we have a function in `node_shared.rb` (which is yet to be modified) and we have a js file for plotting the graph etc.

And voila! We have a chart here!
image

I know this sounds a little complicated at first, but throughout my research in being able to accomplish this, this is the best method I have been able to find, without hampering much of the existing code. The remaining flow is pretty standard, I have just introduced 2-3 extra lines in editor controller and rih.html.erb. Let me know if you follow this!

**Comment by jywarren** [comment]
This looks great; can you actually merge these commits into the PR #5993, so that the tests and functionality are linked into one PR that shows both the functionality and the tests that secure them?

**Comment by IshaGupta18** [comment]
@jywarren we have shiften to https://github.com/publiclab/plots2/pull/6113, which is the latest PR for all the SDG work

**Comment by jywarren** [comment]
OK, great, good to know, and we can close this then, right? THat way people finding this PR aren't confused (as I was! lol)

**Comment by jywarren** [comment]
And likewise should #5993 be closed too and a comment left redirecting to #6113?

**Comment by IshaGupta18** [comment]
Yes, I will be closing that one!

**Pull request closed without merging**